core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala

```diff
@@ -0,0 +1,383 @@
+        activation.andThen {
+            case Success(activation) => sendActiveAck(tid, activation)
+        }.flatMap { activation =>
+            val exec = job.action.exec.asInstanceOf[CodeExec[_]]
+            container.logs(job.action.limits.logs.asMegaBytes, exec.sentinelledLogs).map { logs =>
+                activation.withLogs(ActivationLogs(logs.toVector))
+            }
+        }.andThen {
+            case Success(activation) => storeActivation(tid, activation)
+        }.flatMap { activation =>
+            // Fail the future iff the activation was unsuccessful to facilitate
+            // better cleanup logic.
+            if (activation.response.isSuccess) Future.successful(activation)
+            else Future.failed(ActivationUnsuccessfulError(activation))
+        }
+    }
+}
+
+object ContainerProxy {
+    def props(factory: (TransactionId, String, CodeExec[_], ByteSize) => Future[Container],
+              ack: (TransactionId, WhiskActivation) => Future[Any],
+              store: (TransactionId, WhiskActivation) => Future[Any]) = Props(new ContainerProxy(factory, ack, store))
+
+    // Needs to be threadsafe as it's used by multiple proxys concurrently.
+    private val containerCount = new Counter
+
+    /**
+     * Generates a unique containername
+     *
+     * @param prefix the container name's prefix
+     * @param suffic the container name's suffix
+     * @return a unique container name
+     */
+    def containerName(prefix: String, suffix: String) =
+        s"wsk_${containerCount.next()}_${prefix}_${suffix}".replaceAll("[^a-zA-Z0-9_]", "")
+
+    /**
+     * Creates a WhiskActivation ready to be sent via active ack.
+     *
+     * @param job the job that was executed
+     * @param interval the time it took to execute the job
+     * @param response the response to return to the user
+     * @return a WhiskActivation to be sent to the user
+     */
+    def constructWhiskActivation(job: Run, interval: Interval, response: ActivationResponse) = {
+        val causedBy = if (job.msg.causedBySequence) Parameters("causedBy", "sequence".toJson) else Parameters()
+        WhiskActivation(
+            activationId = job.msg.activationId,
+            namespace = job.msg.activationNamespace,
+            subject = job.msg.user.subject,
+            cause = job.msg.cause,
+            name = job.action.name,
+            version = job.action.version,
+            start = interval.start,
+            end = interval.end,
+            duration = Some(interval.duration.toMillis),
+            response = response,
+            annotations = {
+                Parameters("limits", job.action.limits.toJson) ++
+                    Parameters("path", job.action.fullyQualifiedName(false).toString.toJson) ++ causedBy
+            })
+    }
+}
+
+/** Indicates an activation with a non-successful response */
+case class ActivationUnsuccessfulError(activation: WhiskActivation) extends Exception(activation.toString)
```

Comment:
should this use a more generic exception message rather than serialize the whole record to string?